homebridge/thermostat.ts

```diff
@@ -0,0 +1,178 @@
+              return 'cool'
+            default:
+              return
+          }
+        })()
+        if (!mode) {
+          this.logger.error(
+            `Couldn’t match ${targetHeatingCoolingState} to a recognized mode string.`
+          )
+          return
+        }
+        this.logger.info(`Setting ${this.device.name} mode to ${mode}`)
+
+        return this.device.setInfo({ device: { v1: { mode } } })
+      },
+    })
+
+    this.registerCharacteristic({
+      characteristicType: Characteristic.CurrentTemperature,
+      serviceType: Service.Thermostat,
+      getValue: async () => {
+        const temperature = await this.getTemperatureFromComponentDevices()
+        if (!temperature) {
+          return
+        }
+        // Documentation: https://developers.homebridge.io/#/characteristic/CurrentTemperature
+        // 'Characteristic.CurrentTemperature' supports 0.1 increments
+        return Math.round(temperature * 10) / 10
+      },
+    })
+
+    this.registerCharacteristic({
+      characteristicType: Characteristic.TargetTemperature,
+      serviceType: Service.Thermostat,
+      getValue: ({ setPoint }) => {
+        return setPoint
+      },
+      setValue: (setPoint: number) => {
+        this.logger.info(
+          `Setting ${this.device.name} target temperature to ${setPoint}`
+        )
+
+        // Documentation: https://developers.homebridge.io/#/characteristic/TargetTemperature
+        // 'Characteristic.TargetTemperature' has a valid range from 10 to 38 degrees celsius,
+        // but devices may support a different range. When limits differ, accept the more strict.
+        const setPointMin = Math.max(this.device.data.setPointMin || 10, 10),
+          setPointMax = Math.min(this.device.data.setPointMax || 38, 38)
+
+        if (setPoint < setPointMin || setPoint > setPointMax) {
+          this.logger.error(
+            `Ignoring request to set ${this.device.name} target temperature to ${setPoint}. Target temperature must be between ${setPointMin} and ${setPointMax}.`
+          )
+          return
+        }
+
+        return this.device.setInfo({ device: { v1: { setPoint } } })
+      },
+    })
+
+    this.registerCharacteristic({
+      characteristicType: Characteristic.TemperatureDisplayUnits,
+      serviceType: Service.Thermostat,
+      getValue: () => {
+        // Neither thermostats nor their component devices (e.g. temperature sensors)
+        // appear to include the unit preference. Hardcoding Fahrenheit as the default.
+        return Characteristic.TemperatureDisplayUnits.FAHRENHEIT
+      },
+      setValue: () => {
+        // noop
+        // Setting display unit is unsupported
+      },
+    })
+  }
+
+  async getTemperatureFromComponentDevices(): Promise<number | undefined> {
+    const componentDevices = await this.device.getComponentDevices(),
+      temperatureSensorDevice = componentDevices?.find(({ data }) =>
+        data.deviceType.endsWith('sensor.temperature')
```

Comment:
Just to double-check (because I’m a bit confused by the observables/how updates happen): When a component temperature sensor’s `data` is updated, will this function re-run (and trigger updates to `CurrentTemperature` and `CurrentHeatingCoolingState`, the thermostat characteristics which call it)?

Reply:
I can definitely see where this would be confusing, especially now with component devices in the mix.  Each `RingDevice` has an `onData` observable that will fire whenever the data for that particular device is updated.  Each fires once with the initial data, and then again when we receive websocket messages which are device specific.  `registerCharacteristic` is specifically designed around using the `onData` observable attached to the device used to create the accessory.  This means it will _not_ be listening for changes to the data on the component devices.  So if the temperature sensor updates its data, the parent thermostat will not fire a new event from its `onData`.  In this case, it would be best to grab the child temp sensor `RingDevice` and then map its `onData` observable down to a temperature.  With that observable, you can then use the `registerObservableCharacteristic` instead of `registerCharacteristic`.  Let me know if this is too convoluted, in which case I can just merge what you have and make the changes.

Reply:
> In this case, it would be best to grab the child temp sensor `RingDevice` and then map its `onData` observable down to a temperature. With that observable, you can then use the `registerObservableCharacteristic` instead of `registerCharacteristic`.

I attempted this in https://github.com/dgreif/ring/pull/331/commits/768154bfc9987d2cbc336a21c768fe07c3881b99, but it doesn’t look like the `onValue` methods are getting called, so I’ve missed something. I’ll keep working on it; if you have any suggestions though, please let me know!

Reply:
I think this is the simplest case to illustrate what I don’t understand:
```JS
location.onDevices
.pipe(
  filter((device) => device.initialData.parentZid === this.id || device.data.parentZid === this.id)
)
```
This code does not compile: I’m expecting `device` (`filter` argument) to be a `RingDevice`, but instead it’s a `RingDevice[]`.

Reply:
I think I may have figured it out! I see temperature readings from the component sensors using https://github.com/dgreif/ring/pull/331/commits/e91c6312c4590c735d18964c99fe96a3b81f9cac